### PR TITLE
CryptoPkg: Enable Openssl native instruction support for AARCH64

### PR DESCRIPTION
**Pull request by pierregondois**
# Description
CryptoPkg: Enable Openssl native instruction support for AARCH64

Various OpensslLib implementations are available in edk2. The
OpensslLibAccel.inf and OpensslLibFullAccel.inf ones use
architecture specific instructions, e.g. AESE, PMULL, SHA256H, ...,
allowing to improve speed.

Enable support for Aarch64's native instructions:
- Add ArmReadCntPctReg() and ArmReadIdAA64Isar0Reg() to
  Aarch64's BaseLib.
- Generate Aarch64's specific Openssl functions.
- Add a OpensslStub/AArch64Cap.c file to allow Openssl
  to probe Aarch64 native instruction support.

This patch-set only enable support for GCC for now (MSFT support
not added).

- [ ] Breaking change?
  - **Breaking change** - Does this PR cause a break in build or boot behavior?
  - Examples: Does it add a new library class or move a module to a different repo.
- [x] Impacts security?
  - **Security** - Does this PR have a direct security impact?
  - Examples: Crypto algorithm change or buffer overflow fix.
- [ ] Includes tests?
  - **Tests** - Does this PR include any explicit test code?
  - Examples: Unit tests or integration tests.

## How This Was Tested

The tests run are based on the TestBaseCryptLibShell module.
Each test is run 100 times, then the first 5 values
(considered as warmup) are removed.

The NoAccel column relies on the OpensslLibFull implementation,
the Accel column relies on the OpensslLibFullAccel implementation.
The 'Improvement' column is computed as:
   100 * ('Accel (ns)' - 'NoAccel (ns)') / 'NoAccel (ns)'

The std deviation of the TestVerifyDhGenerateKey is big.
It is due to [1] being called with the 'safe' parameter set, leading
to the prime number taking more time to generate. It requires ~10
iterations when safe=false, ~1000 iterations when safe=true.

The test was run on a Juno-r2. The native Openssl implementation makes
use of the following features (cf. [2]):
- ARMV7_NEON
- ARMV8_AES
- ARMV8_SHA1
- ARMV8_PMULL
- ARMV8_SHA256

and misses:
- ARMV8_SHA512

Note that in order to run the test without exiting too early and actually
see an improvement in the running time, the UT_ASSERT_* macros were bypassed.

See the attached file for the result.
[openssl_native_aarch64.txt](https://github.com/user-attachments/files/16389203/openssl_native_aarch64.txt)

## Integration Instructions

N/A

### CODE REVIEW

**Review by ardbiesheuvel — APPROVED** [review]

**Review by liyi77 — APPROVED** [review]
CryptoPkg part looks good to me.

File: MdePkg/Library/BaseRngLib/AArch64/Rndr.c [hunk 1]
```diff
@@ -43,15 +38,12 @@ BaseRngLibConstructor (
...
-  UINT64  Isar0;
-
   //
   // Determine RNDR support by examining bits 63:60 of the ISAR0 register returned by
   // MSR. A non-zero value indicates that the processor supports the RNDR instruction.
   //
-  Isar0 = ArmReadIdIsar0 ();
-
-  mRndrSupported = ((Isar0 & RNDR_MASK) != 0);
+  mRndrSupported = !((ArmReadIdAA64Isar0Reg () >> ARM_ID_AA64ISAR0_EL1_AES_SHIFT) &
```
Comment (ardbiesheuvel) [inline MdePkg/Library/BaseRngLib/AArch64/Rndr.c:45]: This is using the wrong shift and mask, no?

**Review by ardbiesheuvel — CHANGES_REQUESTED** [review]

File: MdePkg/Library/BaseRngLib/AArch64/Rndr.c (hunk 1 above)
Comment (pierregondois) [reply MdePkg/Library/BaseRngLib/AArch64/Rndr.c:45]: yes indeed ...

File: MdePkg/Library/BaseRngLib/AArch64/Rndr.c [hunk 2]
```diff
@@ -43,15 +38,12 @@ BaseRngLibConstructor (
...
-
   //
   // Determine RNDR support by examining bits 63:60 of the ISAR0 register returned by
   // MSR. A non-zero value indicates that the processor supports the RNDR instruction.
   //
-  Isar0 = ArmReadIdIsar0 ();
-
-  mRndrSupported = ((Isar0 & RNDR_MASK) != 0);
+  mRndrSupported = !((ArmReadIdAA64Isar0Reg () >> ARM_ID_AA64ISAR0_EL1_RNDR_SHIFT) &
+                     ARM_ID_AA64ISAR0_EL1_RNDR_MASK);
```
Comment (ardbiesheuvel) [inline MdePkg/Library/BaseRngLib/AArch64/Rndr.c:46]: Sorry for not spotting this right way, but I think the negation is also wrong.

Shall we just keep the old code instead? I.e.,

```c
Isar0          = ArmReadIdAA64Isar0Reg ();
mRndrSupported = ((Isar0 & ARM_ID_AA64ISAR0_EL1_RNDR_MASK) != 0);
```

**Review by ardbiesheuvel — CHANGES_REQUESTED** [review]

File: MdePkg/Library/BaseRngLib/AArch64/Rndr.c (hunk 2 above)
Comment (pierregondois) [reply MdePkg/Library/BaseRngLib/AArch64/Rndr.c]: My mistake, I should have updated the code to something correct.

File: CryptoPkg/CryptoPkg.dsc [hunk 3]
```diff
@@ -5,6 +5,7 @@
 #  Copyright (c) 2009 - 2022, Intel Corporation. All rights reserved.<BR>
 #  Copyright (c) 2020, Hewlett Packard Enterprise Development LP. All rights reserved.<BR>
 #  Copyright (c) 2022, Loongson Technology Corporation Limited. All rights reserved.<BR>
+#  Copyright (c) 2023, Arm Limited. All rights reserved.<BR>
```
Comment (Wenxing-hou) [inline CryptoPkg/CryptoPkg.dsc:8]: Please use "2024", if there are no special considerations.
There are many places that need to be changed.

File: CryptoPkg/Library/OpensslLib/OpensslLibAccel.inf [hunk 4]
```diff
@@ -5,6 +5,7 @@
 #
```
Comment (Wenxing-hou) [inline CryptoPkg/Library/OpensslLib/OpensslLibAccel.inf:5]: Please change the description too.

This module provides OpenSSL Library implementation with TLS features
along with performance optimized implementations of SHA1, SHA256, SHA512,
AESNI, VPAED, and GHASH for IA32 and X64 and **AARCH64**.

File: CryptoPkg/Library/OpensslLib/OpensslLibFullAccel.inf [hunk 5]
```diff
@@ -10,6 +10,7 @@
 #
```
Comment (Wenxing-hou) [inline CryptoPkg/Library/OpensslLib/OpensslLibFullAccel.inf:10]: Please update the description too.

**Review by ardbiesheuvel — APPROVED** [review]

**Review by Wenxing-hou — APPROVED** [review]
LGTM.

**Pull request merged**